stdlib/Test/src/Test.jl

```diff
@@ -800,37 +800,41 @@ end
 #-----------------------------------------------------------------------
 # Test for log messages
 
-# Test for warning messages (deprecated)
+# Test for output messages (deprecated)
 
-contains_warn(output, s::AbstractString) = occursin(s, output)
-contains_warn(output, s::Regex) = occursin(s, output)
-contains_warn(output, s::Function) = s(output)
-contains_warn(output, S::Union{AbstractArray,Tuple}) = all(s -> contains_warn(output, s), S)
+contains_output(output, s::AbstractString) = occursin(s, output)
+contains_output(output, s::Regex) = occursin(s, output)
+contains_output(output, s::Function) = s(output)
+contains_output(output, S::Union{AbstractArray,Tuple}) = all(s -> contains_output(output, s), S)
 
 """
-    @test_warn msg expr
+    @test_stream stream msg expr
 
-Test whether evaluating `expr` results in [`stderr`](@ref) output that contains
+Test whether evaluating `expr` results in output on the stream that contains
 the `msg` string or matches the `msg` regular expression.  If `msg` is
 a boolean function, tests whether `msg(output)` returns `true`.  If `msg` is a
-tuple or array, checks that the error output contains/matches each item in `msg`.
+tuple or array, checks that the screen output contains/matches each item in `msg`.
 Returns the result of evaluating `expr`.
 
-See also [`@test_nowarn`](@ref) to check for the absence of error output.
-
-Note: Warnings generated by `@warn` cannot be tested with this macro. Use
-[`@test_logs`](@ref) instead.
+[@test_output](@ref) and [@test_warn](@ref) provide utility wrappers for this
+macro for `stdout` and `stderr`.
 """
-macro test_warn(msg, expr)
+macro test_stream(stream, msg, expr)
+    stream === :stdout ? redirect = redirect_stdout :
+        stream === :stderr ? redirect = redirect_stderr :
+        stream === :stdin ? redirect = redirect_stdin :
+        redirect = nothing
     quote
         let fname = tempname()
             try
+                #XXX why not read from the stream directly?
                 ret = open(fname, "w") do f
-                    redirect_stderr(f) do
-                        $(esc(expr))
-                    end
+                    isnothing($redirect) ? $(esc(expr)) :
+                        $redirect(f) do
+                            $(esc(expr))
+                        end
                 end
-                @test contains_warn(read(fname, String), $(esc(msg)))
+                @test contains_output(read(fname, String), $(esc(msg)))
```

Comment:
I copied this directly from the current implementation of `@test_warn`, but am wondering why the macro doesn't just `read()` from the stream directly? Why is the temp file necessary here?